multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/TimeoutAsyncFlowableStep.java

```diff
@@ -50,4 +56,71 @@ private String getStepName() {
     }
 
     public abstract Duration getTimeout(ProcessContext context);
+
+    protected Duration calculateTimeout(ProcessContext context, TimeoutType timeoutType) {
+        Duration timeoutFinal = context.getVariableIfSet(timeoutType.getCommandLineLevelVariable());
+        String timeoutParameterName = timeoutType.getCommandLineAndGlobalLevelParamName();
+
+        if (timeoutFinal == null) {
+            timeoutFinal = extractTimeoutFromAppAttributes(context, timeoutType);
+            if (timeoutFinal != null) {
+                timeoutParameterName = timeoutType.getModuleLevelParamName();
+            }
+        }
+
+        if (timeoutFinal == null) {
+            timeoutFinal = extractTimeoutFromDescriptorParameters(context, timeoutType);
+        }
+
+        if (timeoutFinal == null) {
+            timeoutFinal = Duration.ofSeconds(timeoutType.getCommandLineLevelVariable()
+                                                         .getDefaultValue()
+                                                         .getSeconds());
+            timeoutParameterName = DEFAULT_TIMEOUT;
+        }
+
+        logTimeout(timeoutType.getModuleLevelParamName(), timeoutParameterName, timeoutFinal.toSeconds());
+        return timeoutFinal;
+    }
+
+    private Duration extractTimeoutFromAppAttributes(ProcessContext context, TimeoutType timeoutType) {
+        CloudApplicationExtended app = context.getVariable(Variables.APP_TO_PROCESS);
+        ApplicationAttributes appAttributes = ApplicationAttributes.fromApplication(app, app.getEnv());
+        Number timeout = appAttributes.get(timeoutType.getModuleLevelParamName(), Number.class);
+        if (timeout != null && (timeout.intValue() < 0 || timeout.intValue() > timeoutType.getMaxAllowedValue())) {
```

Comment:
timeoutType.getMaxAllowedValue() is probably a better place for storing the constants for maximum allowed timeouts. Maybe reuse those and set to the correct values in the TimeoutParameterConverter class as well